persistent-postgresql/Database/Persist/Postgresql.hs

```diff
@@ -1738,6 +1761,310 @@ repsertManySql ent n = putManySql' conflictColumns fields ent n
     fields = keyAndEntityFields ent
     conflictColumns = escapeF . fieldDB <$> entityKeyFields ent
 
+-- | This type is used to determine how to update rows using Postgres'
+-- @INSERT ... ON CONFLICT KEY UPDATE@ functionality, exposed via
+-- 'upsertWhere' and 'upsertManyWhere' in this library.
+--
+-- @since 2.12.0.1
+data HandleUpdateCollision record where
+  -- | Copy the field directly from the record.
+  CopyField :: EntityField record typ -> HandleUpdateCollision record
+  -- | Only copy the field if it is not equal to the provided value.
+  CopyUnlessEq :: PersistField typ => EntityField record typ -> typ -> HandleUpdateCollision record
+
+-- | Postgres specific 'upsertWhere'. This method does the following:
+-- It will insert a record if no matching unique key exists.
+-- If a unique key exists, it will update the relevant field with a user-supplied value, however,
+-- it will only do this update on a user-supplied condition.
+-- For example, here's how this method could be called like such:
+--
+-- upsertWhere record [recordField =. newValue] [recordField /= newValue]
+--
+-- Called thusly, this method will insert a new record (if none exists) OR update a recordField with a new value
+-- assuming the condition in the last block is met.
+-- 
+-- @since 2.12.0.1
+upsertWhere
+  :: ( backend ~ PersistEntityBackend record
+     , PersistEntity record
+     , PersistEntityBackend record ~ SqlBackend
+     , MonadIO m
+     , PersistStore backend
+     , BackendCompatible SqlBackend backend
+     )
+  => record
+  -> SqlBackend
+  -> [Update record]
+  -> [Filter record]
+  -> ReaderT backend m ()
+upsertWhere record conn updates filts =
+  upsertManyWhere [record] conn [] updates filts
```

Comment:
I don't know if I love this, since it implies that the user is passing in the `SqlBackend` at call time and I don't like that.  Can I just have this method always use a `postgresql` backend?  What's the pattern for that?

Reply:
So `ReaderT backend` means that we have a `backend` implcity that we can get with `ask :: ReaderT r m r`. So instead of asking for the `conn` as a parameter, we can write:

```haskell
upsertManyWhere records updates filts = do
    conn <- asks projectBackend
    ...
```

(where `projectBackend` uses the `BackendCompatible SqlBackend backend` instance to get the `SqlBackend` out of the `backend`).

Reply:
yeah that's exactly the answer I was looking for, thanks!